File: lemonad-matlab/day15/day15.m
%
% Day 15, Advent of code 2017 (Jonas Nockert / @lemonad)
%
% Note! Takes about 740s to run to completion!
%

tic;

% Example initial values.
% start1 = 65;
% start2 = 8921;

start1 = 516;
start2 = 190;
p = uint64(2147483647);

%
% Part one.
%

N = 40000000;
Gen1 = zeros(N, 1, 'uint32');
Gen2 = zeros(N, 1, 'uint32');
n1 = uint32(start1);
n2 = uint32(start2);
i = 1;
while i <= N
    n1 = uint32(mod(uint64(n1) * 16807, p));
    n2 = uint32(mod(uint64(n2) * 48271, p));
    Gen1(i) = n1;
    Gen2(i) = n2;
    i = i + 1;
end

judges_count = sum(bitand(Gen1, 65535) == bitand(Gen2, 65535));
fprintf('judge''s final count: %d\n', judges_count);


%
% Part two.
%

N = 5000000;
Gen1 = zeros(N, 1, 'uint32');
Gen2 = zeros(N, 1, 'uint32');
n1 = uint32(start1);
n2 = uint32(start2);

% Generator 1 (only pick numbers divisible by 4).
i = 1;
while i <= N
    n1 = uint32(mod(uint64(n1) * 16807, p));
    if ~bitand(n1, 3)
        Gen1(i) = n1;
        i = i + 1;
    end
end

% Generator 2 (only pick numbers divisible by 8).
i = 1;
while i <= N
    n2 = uint32(mod(uint64(n2) * 48271, p));
    if ~bitand(n2, 7)
        Gen2(i) = n2;
        i = i + 1;
    end
end
judges_count = sum(bitand(Gen1, 65535) == bitand(Gen2, 65535));
fprintf('judge''s final count: %d\n', judges_count);

toc;

File: lemonad-matlab/day15/day15.py
"""
Day 15, Advent of code 2017 (Jonas Nockert / @lemonad)

Executes in about five seconds (including both examples from
problem) using pypy.

"""
def next_gen(n, mult, mod=0):
    p = 2147483647
    gen = n
    while True:
        gen = (gen * mult)
        # Modulus with Mersenne Prime (e.g. 2147483647)
        # https://ariya.io/2007/02/modulus-with-mersenne-prime
        # Cuts about 50% execution time, running with pypy.
        x = (gen & p) + (gen >> 31)
        gen = (x - p) if x >= p else x
        if not gen & mod:
            yield gen & 65535

def compare(gen1, gen2, N):
    counter = 0
    for i in range(0, N):
        if next(gen1) == next(gen2):
            counter = counter + 1
    return counter


#
# Part 1
#

N = 40000000
gen1 = next_gen(516, 16807)
gen2 = next_gen(190, 48271)
counter = compare(gen1, gen2, N)
print("Judge's final count (part one): %d" % counter)
assert(counter == 597)

# Check example from problem.
gen1 = next_gen(65, 16807)
gen2 = next_gen(8921, 48271)
counter = compare(gen1, gen2, N)
assert(counter == 588)


#
# Part 2
#

N = 5000000
gen1 = next_gen(516, 16807, 3)
gen2 = next_gen(190, 48271, 7)
counter = compare(gen1, gen2, N)
print("Judge's final count (part two): %d" % counter)
assert(counter == 303)

# Check example from problem.
gen1 = next_gen(65, 16807, 3)
gen2 = next_gen(8921, 48271, 7)
counter = compare(gen1, gen2, N)
assert(counter == 309)
